src/scanner.cpp

```diff
@@ -0,0 +1,213 @@
+{
+  while (peek() != '"' && !end_reached()) {
+    if (peek() == '\n') { ++line; }
+    advance();
+  }
+
+  if (end_reached()) {
+    report(line, "Unterminated string");
+    return;
+  }
+
+  advance();
+
+  std::string_view value = source.substr(start + 1, current - start - 2);
+  add_token(TokenType::STRING, std::string{ value });
+}
+
+void Scanner::scan_number()
+{
+  while (util::is_alphanumeric(peek()) || peek() == '.') { advance(); }
+  double val = NAN;
+  std::string_view str_val = source.substr(start, current - start);
+  auto [ptr, ec] = std::from_chars(str_val.data(), str_val.data() + str_val.size(), val);
+  if (ec == std::errc() && ptr == str_val.data() + str_val.size()) {
+    add_token(TokenType::NUMBER, val);
+  } else {
+    report(line, "Invalid number: " + std::string{ str_val });
+  }
+}
+
+void Scanner::scan_identifier()
+{
+  while (util::is_alphanumeric(peek())) { advance(); }
+  std::string_view lexeme = source.substr(start, current - start);
+  auto it = ::keywords.find(lexeme);
+  if (it != ::keywords.end()) {
+    add_token(it->second);
+  } else {
+    add_token(TokenType::IDENTIFIER);
+  }
+}
+
+void Scanner::scan_token()
+{
+  char c = advance();
+  switch (c) {
+  case '(':
+    add_token(TokenType::LEFT_PAREN);
+    break;
+  case ')':
+    add_token(TokenType::RIGHT_PAREN);
+    break;
+  case '{':
+    add_token(TokenType::LEFT_BRACE);
+    break;
+  case '}':
+    add_token(TokenType::RIGHT_BRACE);
+    break;
+  case ',':
+    add_token(TokenType::COMMA);
+    break;
+  case '.':
+    add_token(TokenType::DOT);
+    break;
+  case '-':
+    add_token(TokenType::MINUS);
+    break;
+  case '+':
+    add_token(TokenType::PLUS);
+    break;
+  case ';':
+    add_token(TokenType::SEMICOLON);
+    break;
+  case '*':
+    add_token(TokenType::STAR);
+    break;
+  case '!':
+    add_token(match('=') ? TokenType::BANG_EQUAL : TokenType::BANG);
+    break;
+  case '=':
+    add_token(match('=') ? TokenType::EQUAL_EQUAL : TokenType::EQUAL);
+    break;
+  case '<':
+    add_token(match('=') ? TokenType::LESS_EQUAL : TokenType::LESS);
+    break;
+  case '>':
+    add_token(match('=') ? TokenType::GREATER_EQUAL : TokenType::GREATER);
+    break;
+  case '/':
+    if (match('/')) {
+      while (peek() != '\n' && !end_reached()) { advance(); }
+    } else {
+      add_token(TokenType::SLASH);
+    }
+    break;
+
+  case ' ':
+  case '\r':
+  case '\t':
+    break;
+
+  case '\n':
+    line++;
+    break;
+
+  case '"':
+    scan_string();
+    break;
+  default:
+    if (util::is_digit(c)) {
+      scan_number();
+    } else if (util::is_alpha(c)) {
+      scan_identifier();
+    } else {
+      report(line, "Unexpected character: " + std::string{ c });
+    }
+  }
+}
+
+[[nodiscard]] ParsedTokens Scanner::scan()
+{
+  while (!end_reached()) {
+    start = current;
+    scan_token();
+  }
+
+  tokens.push_back(Token{ .line = line, .type = TokenType::END_OF_FILE, .lexeme = "", .literal = std::nullopt });
```

Comment:
If `.literal` is `std::nullopt` (or just `{}`) by default, no need to assign it here, that's half the benefit of using designated init: you can skip stuff you don't need to override the defaults of.